src/boot/efi/util.c

```diff
@@ -769,6 +770,107 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         return EFI_SUCCESS;
 }
 
+#define QEMU_KERNEL_LOADER_FS_MEDIA_GUID \
+        &(const EFI_GUID) { 0x1428f772, 0xb64a, 0x441e, {0xb8, 0xc3, 0x9e, 0xbd, 0xd7, 0xf8, 0x93, 0xc7 }}
+
+#define QEMU_BOOT_ORDER_GUID \
+        &(const EFI_GUID) { 0x668f4529, 0x63d0, 0x4bb5, {0xb6, 0x5d, 0x6f, 0xbb, 0x9d, 0x36, 0xa4, 0x4a }}
+
+/* detect direct boot */
+bool is_qemu_direct_boot(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        /* 'qemu -kernel systemd-bootx64.efi' */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_VENDOR_DP &&
+            memcmp(&dp->Guid, QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) == 0)
+                return true;
+
+        /* loaded from firmware volume (sd-boot added to ovmf) */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_PIWG_FW_VOL_DP)
+                return true;
+
+        return false;
+}
+
+/* return the length of the device path in bytes, excluding the end tag */
+static size_t device_path_length(EFI_DEVICE_PATH *dp) {
+        size_t len;
+
+        for (len = 0; !IsDevicePathEnd(dp); dp = NextDevicePathNode(dp))
+                len += DevicePathNodeLength(dp);
+        return len;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *ret_qemu_dev, EFI_FILE **ret_qemu_dir) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        UINTN n_handles;
+        EFI_STATUS err;
+
+        /* ovmf doesn't connect devices for direct kernel boot */
+        reconnect();
+
+        /* find all file system handles */
+        err = BS->LocateHandleBuffer(ByProtocol, &FileSystemProtocol, NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        for (UINTN order = 0;; order++) {
+                _cleanup_free_ EFI_DEVICE_PATH *dp = NULL;
+                char16_t order_str[20];
+                size_t dp_len = 0;
+
+                SPrint(order_str, sizeof(order_str), L"QemuBootOrder%04x", order);
```

Comment:
Looks like this is brand new stuff, no?

In https://github.com/tianocore/edk2/blob/b7213bbd59833fb0786c83a28df5f8244602ab5e/OvmfPkg/Library/QemuBootOrderLib/QemuBootOrderLib.c#L1797 it uses %d for the varname. That doesn't sound intentional to me.

Reply:
Yep, it's new, you effectively need git master to try that, an incremental fix was committed just this week.
And, yes, it should be fixed to use %x not %d for consistency with BootNNNN, DriverNNNN etc.